### ISSUE TITLE
pkg_resources.DistributionNotFound:

### ISSUE THREAD

**Issue by Roy-Kid** [jewettaij/moltemplate#36]
Dear Author：

An error occured. I used "pip3 install moltemplate" to install latest version, then I typed "moltemplate.sh system.lt" and it told me that :

moltemplate.sh v2.17.10 2020-6-12

Traceback (most recent call last):
  File "/home/roy/.local/bin/lttree_check.py", line 6, in <module>
    from pkg_resources import load_entry_point
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 3088, in <module>
    @_call_aside
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 3072, in _call_aside
    f(*args, **kwargs)
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 3101, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 574, in _build_master
    ws.require(__requires__)
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 892, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/usr/lib/python2.7/dist-packages/pkg_resources/__init__.py", line 778, in resolve
    raise DistributionNotFound(req, requirers)
pkg_resources.DistributionNotFound: The 'moltemplate==2.17.10' distribution was not found and is required by the application.

And i tried sudo pip3 but also failed.

When i git clone the source code from github, and use both "pip3 install . " and "pip3 install --user .", but:

moltemplate.sh v2.17.10 2020-6-12

Traceback (most recent call last):
  File "/home/roy/.local/bin/lttree_check.py", line 5, in <module>
    from moltemplate.lttree_check import main
ImportError: No module named moltemplate.lttree_check

So i dont know what's wrong. I will try old version to check if those mistakes is owe to my fault.

**Comment by jewettaij** [comment]
I could be wrong, but at first glance, this really looks like a clash between different versions of python installed on your system.  I say this because you used pip3 to install moltemplate, however when you run moltemplate.sh, it is generating error messages (traceback messages) which refer to python2.7.  ("File "/usr/lib/python2.7/dist-packages/pkg_resources/init.py", line 3088, in").  
So, before doing anything else, uninstall moltemplate using:
```
pip3 uninstall moltemplate
```

Suggestion 1: What happens if you just use "pip" instead of "pip3"?

Suggestion 2: You might have a broken or corrupted python installation.  Try deleting python and reinstalling it.
(Alternatively, you can override your existing python by installing the "anaconda" version of python on your computer.  I've had a good experience using the anaconda or miniconda distributions of python.  Perhaps you are using these already.)

I'll wait until I hear back from you before closing this issue.

P.S.
(Next time you install moltemplate, I suggest doing it in a virtual environment.  IE use "python -m venv venv" and then "source venv/bin/activate" before running "pip install moltemplate".)
```
python -m venv venv  # <-- OPTIONAL
source venv/bin/activate   # <-- OPTIONAL
pip install moltemplate
```

**Comment by Roy-Kid** [comment]
Dear sir：
Thank you for your suggestion！I have solved this problem by using a “pipenv” virtual environment. But I still don't know what's wrong. I have used moltemplate this brilliant modeling software for two or three years. But after a half year interval, the moltemplate on my computers all go wrong, not only my laptop but also my desktop and HPC. It's weird. 

First, for suggestion1, there is two python version on my computer, 2.7 and 3.8. Yes I have tied 
```
pip install moltemplate
pip3 install moltemplate
python -m pip install moltemplate
python3 -m pip install moltemplate
```
but all not work, same as "pip install ." in moltemplate directory. 

Second, I have not tried reinstalling my python and I dont think it will work. Because all my softwares and other packages depend on python still work. If I have spare time i would have a try on another new environment. 

And, my friend told me yesterday on his laptop also meet the same problem but his HPC is ok. 

Not a big deal, we can fix it by virtual env!

(I dont know the courtesy to the Github issue, so please you determine when to close this issue~)

**Comment by jewettaij** [comment]
Thank you for replying.  Lots of helpful information there.  I still have a few more questions and suggestions about the computers you and your friend are having difficulty installing moltemplate on:

### Questions
1) What operating systems do they run?  (For example: MacOS Sierra, MacOS Catalina, Ubuntu linux, CentOS linux, Mint Linux, Amazon Linux, Windows10 WSL, Windows 10 HyperV, Windows10+VirtualBox+Ubuntu?)
2) What happens when you type
```
which python
which python3
which pip
which pip3
```
I'm trying to figure out whether you are using the version of python which comes with these operating systems by default?  Or did you reinstall python?  Some programs (like the "Chimera" visualization tool) install their own custom version of python and override the user's PATH so that this version of python gets used instead.  I want to see if  the "pip" and "python" that you are using are from the same python distribution.

### previous issues

A couple months ago a couple users complained that moltemplate was not working when installed using pip.  It turned out there was a clash between the different versions of python installed on their computer.  The fact that you are getting an error message that includes "python2.7" makes me think something similar is happening to you.  I'm not sure why I'm getting more complaints about this now than in the past.

For example, sometimes a user installs anaconda python, but the "pdb" (and "pip"?) programs they are using are for their old python that came with the computer's operatying system (not from anaconda).  There are so many weird things that can happen when there are multiple versions of python present.  So I really want to rule this out.

**Comment by Roy-Kid** [comment]
# Answers
1. We are all run Ubuntu18.04 on our computer, both original and WSL(1 & 2 both). We have no custom but all default configurations. 

2. 
```
roy@Carbon:~$ which python                                                                /usr/bin/python                                                                           roy@Carbon:~$ which python3                                                               /usr/local/bin/python3                                                                    roy@Carbon:~$ which pip                                                                   /home/roy/.local/bin/pip                                                                  roy@Carbon:~$ which pip3                                                                  /home/roy/.local/bin/pip3
```
No, I have updated python3 to 3.8 but python2 is the default. This update has been a long time, since then still a long time that moltemplate works well. I think maybe the COVID-19 makes this software break down XD。 

3. I have no idea whether a software overrides my PATH because I am not an expert to Linux, I just follow those manuals. 
```
# $PATH on my WSL 
-bash: /home/roy/.local/bin:/home/roy/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/mnt/c/WINDOWS/system32:/mnt/c/WINDOWS:/mnt/c/WINDOWS/System32/Wbem:/mnt/c/WINDOWS/System32/WindowsPowerShell/v1.0/:/mnt/c/WINDOWS/System32/OpenSSH/:/mnt/c/Program: No such file or directory 
```

4. Yes, it might be related to versions of python and the system, because when I use a virtual environment then all clear.

5. Another issue: I sincerely hope you can add the content to the manual, which explains how to use -pdb and -xyz flags in command line to convert .pdb and .xyz file to .lt. My friend and I can not figure out how to do it. Thanks!

**Comment by jewettaij** [comment]
> # Answers
> 1. We are all run Ubuntu18.04 on our computer, both original and WSL(1 & 2 both). We have no custom but all default configurations.
> 
> ```
> roy@Carbon:~$ which python                                                                /usr/bin/python                                                                           roy@Carbon:~$ which python3                                                               /usr/local/bin/python3                                                                    roy@Carbon:~$ which pip                                                                   /home/roy/.local/bin/pip                                                                  roy@Carbon:~$ which pip3                                                                  /home/roy/.local/bin/pip3
> ```

I see that you installed "pip" and "pip3" yourself locally (in your /home/roy/.local/ directory).  However python and python3 were installed system-wide by the administrator (in /usr/bin/ and /user/local/bin/). 
 Perhaps the version of pip (or pip3) you installed is not compatible with the version of python you have on your system.

Is there a version of pip in /usr/bin/ ?
Is there a version of pip3 in /usr/local/bin/ ?
If so, use that version of pip or pip3 (instead of the version of pip in /home/roy/.local/)

I like to install anaconda locally because it installs python, numpy, pip, wheel, and many other packages automatically.  This makes sure they are all compatible.  Typically I install anaconda locally (for example in /home/roy).

> No, I have updated python3 to 3.8 but python2 is the default. This update has been a long time, since then still a long time that moltemplate works well. I think maybe the COVID-19 makes this software break down XD。

I hope not.

> 1. I have no idea whether a software overrides my PATH because I am not an expert to Linux, I just follow those manuals.
> 
> ```
> # $PATH on my WSL 
> -bash: /home/roy/.local/bin:/home/roy/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/mnt/c/WINDOWS/system32:/mnt/c/WINDOWS:/mnt/c/WINDOWS/System32/Wbem:/mnt/c/WINDOWS/System32/WindowsPowerShell/v1.0/:/mnt/c/WINDOWS/System32/OpenSSH/:/mnt/c/Program: No such file or directory 
> ```

This looks okay.

> 1. Yes, it might be related to versions of python and the system, because when I use a virtual environment then all clear.

Okay.  Perhaps this is the solution.  Perhaps I'll recommend doing it this way on the README.md file.

> 2. Another issue: I sincerely hope you can add the content to the manual, which explains how to use -pdb and -xyz flags in command line to convert .pdb and .xyz file to .lt. My friend and I can not figure out how to do it. Thanks!

I hope I am not misleading everyone.
Moltemplate cannot automatically convert PDB and XYZ files into LAMMPS files.
Moltemplate only reads the X,Y,Z coordinates from these files.  It is the responsibility of the user to supply the bonds and force-field information.
If you think I should clarify this in the manual, I will.

Chapter 4 of the moltemplate manual has an example explaining how to read coordinates from a PDB.
There's also a detailed example here:

https://github.com/jewettaij/moltemplate/tree/master/examples/file_conversion_examples/read_PDB_file_examples/waterSPCE_from_PDBfile

However, you will notice that in these examples, the user is required to specify all of the molecular details, including bond topology. (In this case the "spce.lt" file contains the details for the water molecule).  Moltemplate cannot infer topology or force-field information from a PDB file (or XYZ file).  Perhaps I should make this more clear in the manual.  Some programs can do this, although they often fail.  Moltemplate does not even attempt to do this.

Typically most molecule builders attempt to assign bond, angle, dihedral, and improper force field parameters according to the types of the atoms in a way that depends on the rules of the force field.  However force fields typically have many different kinds of carbon atoms, as well as many types of oxygen, hydrogen, sulfer, atom types.  PDB and XYZ files contain the element of each atom (for example, carbon, hydrogen, oxygen,...), but they do not specify which kind of carbon, hydrogen, or oxygen atom it is.  Some molecule-builder programs have the ability to automatically infer "atom typing".  In other words: These programs can attempt infer bond connectivity from PDB files.   Once the bonds are determined, these programs try to figure out which type of carbon atom (for example) should be used at each location.  These tools are very useful, but they usually only work with one type of force field.  Moltemplate will probably never do this.

Again, if you think I should clarify this in the manual or the web page, I am happy to do this.

Andrew

**Comment by Roy-Kid** [comment]
Thanks for your further explanation to my question! I want to use packmol to filling my system with water, but I find it will output a pdb data file. So I need to use moltemplate to "convert" this file to lammps data. 

So I just come back to read again the chapter about SPCE water in manual. What I think is:
1. give a .lt file to specify A molecule topology(atoms coordinate，bond connectivity and so on)
2. In system.lt we specify the system information, such as the number of the molecules, boundaries.
3. Then moltemplate use the coordinates in pdb to guide how to copy molecules in the system.

OK, I think I know what it can not convert to lammps data directly! Moltemplate can not map the atom type in pdb to type in forcefield file(such as "oplsaa.lt"). So we should point out by using (@atom: ). That means once we give full information to a molecule, then moltemplate will know how to contribute forcefield, then just use pdb to copy and move those molecules?

That is what I comprehend the manual. I really hope you can clarify this -pdb flag, hence most of modeling Softwares are support .pdb file but very file to lammps data. If you can improve this function, then can set up a bridge from main Softwares to LAMMPS, greatly reduce the difficulty to use lammps. (I think the most difficult to use lammps is modeling, then find a forcefield. Use that software is really easy...  if I can know how to use packmol and moltemplate -pdb a little bit earlier, I can save two days to model a lysine by hand)

Thanks for your selfless contribution to lammps and open-source software, I would learn this sprite from you. This day i set up a lammps tutorial [website](https://lammps.org.cn) for how to use lammps and moltemplate, I think if we can sort out a tutorial for those starters, they can save a lot of time in deal with those problems and do what they should do.

**Comment by jewettaij** [comment]
> Thanks for your further explanation to my question! I want to use packmol to filling my system with water, but I find it will output a pdb data file. So I need to use moltemplate to "convert" this file to lammps data.
> 
> So I just come back to read again the chapter about SPCE water in manual. What I think is:
> 
> 1. give a .lt file to specify A molecule topology(atoms coordinate，bond connectivity and so on)
> 2. In system.lt we specify the system information, such as the number of the molecules, boundaries.
> 3. Then moltemplate use the coordinates in pdb to guide how to copy molecules in the system.
> 
> OK, I think I know what it can not convert to lammps data directly! Moltemplate can not map the atom type in pdb to type in forcefield file(such as "oplsaa.lt"). So we should point out by using (@atom: ). That means once we give full information to a molecule, then moltemplate will know how to contribute forcefield, then just use pdb to copy and move those molecules?

This is exactly correct.

> That is what I comprehend the manual. I really hope you can clarify this -pdb flag, hence most of modeling Softwares are support .pdb file but very file to lammps data.

The next version of the manual will explain this more carefully.

> If you can improve this function, then can set up a bridge from main Softwares to LAMMPS, greatly reduce the difficulty to use lammps.

I agree that this feature could be improved.  But I am curious to know what you think.  Do you have a suggestion to improve PDB file reading?

> (I think the most difficult to use lammps is modeling, then find a forcefield. Use that software is really easy... if I can know how to use packmol and moltemplate -pdb a little bit earlier, I can save two days to model a lysine by hand)

For a small organic molecule like lysine, I have been manually copying the X,Y,Z coordinates from the PDB file to the write("Data Atoms") section, one atom at a time.  It would be better if moltemplate could read the coordinates for a single molecule from a PDB file.  Currently moltemplate can only read the PDB file for the entire system (containing all of the molecules).

Are you trying to use moltemplate to prepare a simulation containing a protein?

If so, this requires effort.  I am not aware of anybody who has used moltemplate to simulate proteins.  (I suspect that somebody would have to create LT files for each of the 20 amino acids, as well as any post-translationally modified amino acids.)

a) Use PACKMOL to create a new XYZ file or PDB file containing the protein and the water molecules.  (I prefer XYZ files.)
b) Create a system.lt file containing the topology for the entire system including the protein and the water.  (This file will probably refer to other files which define the amino acids.)
c) Run:
```
moltemplate -xyz packmol_file.xyz system.lt
```
The atoms should be in the same order they appear in the XYZ file or PDB file.  (Depending on the PDB file or XYZ file, the order could be very different.  If this is the issue causing you the most trouble, we can discuss how to solve this problem.)

> Thanks for your selfless contribution to lammps and open-source software, I would learn this sprite from you.

You're welcome and thanks for the compliment.

> This day i set up a lammps tutorial [website](https://lammps.org.cn) for how to use lammps and moltemplate,

Cool!!

> I think if we can sort out a tutorial for those starters, they can save a lot of time in deal with those problems and do what they should do.

Creating good documentation and tutorials is hard work.  The kinds of simulations that I run are coarse-grained simulations.  I do not have much experience running "all-atom" simulations.  So feedback from you is useful to me.  I'm also glad you are creating your own documentation as well.  Thanks for doing this.

Andrew

**Comment by Roy-Kid** [comment]
Sir：
I have figure out how to use -pdb and -xyz to build a system, thanks for your kindly help! 

There are many examples in the moltemplate file that I omitted. It is a little bit inconspicuous; next time you update your manual, you can mention this at the first of the tutorial. Then tell users how to use those two flags~. I have written a simple tutorial on my website. My website is multilanguage support. If you think it is necessary, we can translate it to English as a fundamental tutorial to use moltemplate.

These days I have built a C16-lysine-lysine molecule in OPLSAA with TIP4P water. If you think it is a  synthetical example，i am glad to email those lt files to post on moltemplate website.

And my friend, who studied at Johns Hopkins University as a postgraduate, he plans to write 20 amino acid lt moltemplate. We think after a proper arrangement, we can build peptide and small protein. If we have any progress, I would email to you. 

Thanks again for your selfless contribution! 

Best Wish

Roy Kid 
University of Science & Technology of China

**Comment by Roy-Kid** [comment]
I find another problem:

I use ChemDraw to draw a pdb of lysine, then write a .lt by hand. Then I find that the order of atoms in pdb do not match the order in my .lt file. That means although every position does have an atom, not the right one:
for example: 
the order in pdb is H-H-O (0,0,0) (1,0,0) (2,0,0)
the order in system.lt is O-H-H (0,0,0) (1,0,0) (2,0,0)
the order in system.data is O-H-H (0,0,0) (1,0,0) (2,0,0)

I think I ask too much for this function, but it does a problem.
well, the question is, how to convert lt to pdb or from pdb to lt?

Sorry to bother you

**Comment by jewettaij** [comment]
> Sir：
> I have figure out how to use -pdb and -xyz to build a system, thanks for your kindly help!
> 
> There are many examples in the moltemplate file that I omitted. It is a little bit inconspicuous; next time you update your manual, you can mention this at the first of the tutorial. Then tell users how to use those two flags~.

Next time I update the moltemplate manual, I will include a larger explanation of the "-pdb" and "-xyz" arguments, and I will also mention their existence at the beginning of the first tutorial (chapter 4 of the manual).  Thank you for the suggestion.

> I have written a simple tutorial on my website. My website is multilanguage support. If you think it is necessary, we can translate it to English as a fundamental tutorial to use moltemplate.

I am very flattered that you included moltemplate information on your web site.  I took a quick look.  I don't understand Chinese, but I used the "google translate" service to read the first half of your page and the information seems good.  (See "google_translate_moltemplate.txt", which I attached.)  Tomorrow I will take another look.  I used an automatic translation tool to convert the text to English  Sometime I will read this file more carefully and edit i
[moltemplate_tools_tutorial_Roy_Kid_2020-7-17.txt](https://github.com/jewettaij/moltemplate/files/4940028/moltemplate_tools_tutorial_Roy_Kid_2020-7-17.txt)
t on github.  It seems like a valuable resource!

> These days I have built a C16-lysine-lysine molecule in OPLSAA with TIP4P water. If you think it is a synthetical example，i am glad to email those lt files to post on moltemplate website.

Sure!

> And my friend, who studied at Johns Hopkins University as a postgraduate, he plans to write 20 amino acid lt moltemplate. We think after a proper arrangement, we can build peptide and small protein. If we have any progress, I would email to you.

That would be wonderful.  If he creates them, I will post them.  But I warn you that, more work will be required to make it easy to prepare simulations of proteins using moltemplate.  I think that most people who simulate proteins do not use LAMMPS.  Programs like AMBER, GROMACS, NAMD, CHARMM, and OpenMM (which is a good program) come with molecule builder programs which were designed to create proteins in water.  (These tools are not usually very good at some other tasks, such as coarse-grained modeling, but they are very good at reading PDB files and preparing simulations of proteins in water.) Some LAMMPS users create their simulation input files using "AmberTools" as well as a conversion program (like amber2lmp) which converts AMBER input files into LAMMPS format.  Even with the LT files for the 20 amino acids, moltemplate will not be as advanced as these tools.
> Thanks again for your selfless contribution!

Thank you for yours!

> Best Wish
> 
> Roy Kid
> University of Science & Technology of China

**Comment by jewettaij** [comment]
> I find another problem:
> 
> I use ChemDraw to draw a pdb of lysine, then write a .lt by hand. Then I find that the order of atoms in pdb do not match the order in my .lt file. That means although every position does have an atom, not the right one:
> for example:
> the order in pdb is H-H-O (0,0,0) (1,0,0) (2,0,0)
> the order in system.lt is O-H-H (0,0,0) (1,0,0) (2,0,0)
> the order in system.data is O-H-H (0,0,0) (1,0,0) (2,0,0)
> 
> I think I ask too much for this function, but it does a problem.
> well, the question is, how to convert lt to pdb or from pdb to lt?

Two ways:
1) Change the order of the lines in the "Data Atoms" section of the LT file to match the PDB file.
Old:
```
  write("Data Atoms") {
    $atom:o    $mol    @atom:O -0.8476  0.0000000 0.00000 0.000000
    $atom:h1   $mol    @atom:H  0.4238  0.8164904 0.00000  0.5773590
    $atom:h2   $mol    @atom:H  0.4238  -0.8164904 0.00000 0.5773590
  }
```
New:
```
  write("Data Atoms") {
    $atom:h1   $mol    @atom:H  0.4238  0.8164904 0.00000  0.5773590
    $atom:h2   $mol    @atom:H  0.4238  -0.8164904 0.00000 0.5773590
    $atom:o    $mol    @atom:O -0.8476  0.0000000 0.00000 0.000000
  }
```
2) Change the order of the lines in the PDB file created by 
```
ATOM     64  H1  TIP3W 150       3.303  17.803   7.322  1.00  0.00      WT1  H
ATOM     65  H2  TIP3W 150       4.533  18.755   6.924  1.00  0.00      WT1  H
ATOM     66  OH2 TIP3W 150       4.299  17.774   7.169  1.00  0.00      WT1  O
```
New:
```
ATOM     66  OH2 TIP3W 150       4.299  17.774   7.169  1.00  0.00      WT1  O
ATOM     64  H1  TIP3W 150       3.303  17.803   7.322  1.00  0.00      WT1  H
ATOM     65  H2  TIP3W 150       4.533  18.755   6.924  1.00  0.00      WT1  H
```
I don't think moltemplate cares about the numbers in the second column (64,65,66).  You do not have to change them.

Perhaps I should implement a feature which allows moltemplate to interpret the name of each atom in the PDB file (eg. "OH2", "H1", "H2"), so that it will attempt to match with the names of the $atom variables in the "Data Atoms" section.  I probably won't have time to add this feature soon, but perhaps I can do it by the end of the year.

It's exciting that you are interested in creating LT files for the 20 amino acids.  I just worry that you will still be disappointed afterwards, because there is still more work to do.

> Sorry to bother you

I hope this helps.

Andrew

**Comment by Roy-Kid** [comment]
> It's exciting that you are interested in creating LT files for the 20 amino acids. I just worry that you will still be disappointed afterwards, because there is still more work to do.

After I have find that it must match the order of .pdb and .lt file, i give up. I write a python file that can convert .pdb to .lt, but in the vanilla i left "atom type" column blank. i try to find a way to parse forcefield.lt in order to automatic match pdb atom type in its 13-16 column to the @atom(or biotype) in oplsaa.lt(for example). 

Changing the order of .lt or .pdb is impossible there is too many atoms in a molecule. So the only way is generate a .lt file from  .pdb. (i think)

> I am very flattered that you included moltemplate information on your web site. I took a quick look. I don't understand Chinese, but I used the "google translate" service to read the first half of your page and the information seems good. (See "google_translate_moltemplate.txt", which I attached.) Tomorrow I will take another look. 

The fundamental tutorial is written when i was a sophomore, it is really simple. I would like to give a more specific one and translate it english. (you can find a select language drop-down meun at the nav bar). 

>  I think that most people who simulate proteins do not use LAMMPS. Programs like AMBER, GROMACS, NAMD, CHARMM, and OpenMM (which is a good program) come with molecule builder programs which were designed to create proteins in water. 

My question is how are they better than LAMMPS(except for some are commercial softwares, i only consider open source software, NAMD OpenMM and gromacs), just because they have supporting modeling tools? Or they are fast or accurate? Or they have embed functions to calculate more parameters.

PS: there is a little problem in moltemplatee's example: at all_atom/force_field_oplsaa/waterSPC_using_oplsaa, there is no oplsaa.lt file in this folder. Consider that you have provide out-of-box-like .sh script, there is no reason that let the starter to find oplsaa.lt in moltemplate/moltemplate/force_fields.

I hope we can get together to make moltemplate great XD.

**Comment by jewettaij** [comment]
I just wanted to let you know that I am too busy to work on this in the next week.  But I will reply to this next week.  (If I forget, please email me.)

**Comment by jewettaij** [comment]
> > It's exciting that you are interested in creating LT files for the 20 amino acids. I just worry that you will still be disappointed afterwards, because there is still more work to do.
> 
> After I have find that it must match the order of .pdb and .lt file, i give up. I write a python file that can convert .pdb to .lt, but in the vanilla i left "atom type" column blank. i try to find a way to parse forcefield.lt in order to automatic match pdb atom type in its 13-16 column to the @atom(or biotype) in oplsaa.lt(for example).
> 
> Changing the order of .lt or .pdb is impossible there is too many atoms in a molecule. So the only way is generate a .lt file from .pdb. (i think)
> 
> > I am very flattered that you included moltemplate information on your web site. I took a quick look. I don't understand Chinese, but I used the "google translate" service to read the first half of your page and the information seems good. (See "google_translate_moltemplate.txt", which I attached.) Tomorrow I will take another look.
> 
> The fundamental tutorial is written when i was a sophomore, it is really simple. I would like to give a more specific one and translate it english. (you can find a select language drop-down meun at the nav bar).
> 
> > I think that most people who simulate proteins do not use LAMMPS. Programs like AMBER, GROMACS, NAMD, CHARMM, and OpenMM (which is a good program) come with molecule builder programs which were designed to create proteins in water.
> 
> My question is how are they better than LAMMPS(except for some are commercial softwares, i only consider open source software, NAMD OpenMM and gromacs), just because they have supporting modeling tools? Or they are fast or accurate? Or they have embed functions to calculate more parameters.

Molecule building programs like AmberTools, CharmmGUI, Chimera, MosDeF/Foyer/Mbuild, ParmEd, (and VMD+plugins), can do many things which moltemplate cannot do:

1) Understand PDB files amd mCIF files.
a) Add bonds between nearby atoms automatically.
b) Repair incomplete in PDB files, such as adding hydrogen atoms or adding side-chains.  (Many PDB files only contain backbone atoms or heavy atoms.)
c) Calculate the charge of each atom using quantum-chemistry methods

2) "Atom-typing":  Automatically determine atom type from bond connectivity.  It is not difficult to create a software program that can recognize the names of atoms in PDB files which belong to the the standard 20 amino acids "CA", "CB", "O", "H11", "H12", ... and choose the atom which is appropriate for the force-field.  (For example, when using OPLSAA, the "CA" alpha-carbon atom corresponds to @atom:166 in "oplsaa.lt", and either "@atom:c3", or "@atom:cx" or "@atom:cy" in "gaff2.lt".  I'm not certain, and I'm too)

  For a detailed example, see this web page:
http://ambermd.org/tutorials/basic/tutorial5/

  It is difficult to write a general tool which can guess the best (most similar) atom type for other exotic molecules, such as ligands and some post-translationally modified amino acids.  Even the programs listed above often have difficulty solving this problem.

  Moltemplate does not attempt to solve problems 1 or 2.  Moltemplate does not choose atom type.  Users currently must carefully specify exactly which atom type to use in every molecule (for example, "@atom:166", "@atom:c3", "@atom:cx", "@atom:cy").  This is a lot of work, and if the user makes a mistake, their simulations will be incorrect.  I am honestly quite worried that people will use moltemplate to prepare simulations using the wrong atom types and they will get the wrong results.  There is a big warning on the moltemplate web page, but I worry that people do not notice it:
http://moltemplate.org/force_field_recommendations.html

   Moltemplate was intended to be used for coarse-grained simulations.  I will probably never write code that solves problems 1 and 2 for other force fields, like AMBER/GAFF or OPLSAA.   It is too much work for me to write tools that can solve problems 1) and 2) above, for every possible type of PDB file and force field.  This is why for proteins I currently recommend using one of the tools above to prepare simulations with complex molecules (like proteins), instead of using moltemplate.  Once you have files prepared for AMBER, or CHARMM, (or in the future OpenMM), it should be possible to convert those files to LAMMPS format.  

Currently, I do not recommend using moltemplate to prepare simulations of complex molecules like proteins.  Instead I would use one of the molecule builder programs above, and convert its files into LAMMPS format.  Then if you want to use moltemplate, you can convert the LAMMPS data file containing the protein into an .LT file using the "ltemplify.py" converter program (which is included with moltemplate).

> PS: there is a little problem in moltemplatee's example: at all_atom/force_field_oplsaa/waterSPC_using_oplsaa, there is no oplsaa.lt file in this folder. Consider that you have provide out-of-box-like .sh script, there is no reason that let the starter to find oplsaa.lt in moltemplate/moltemplate/force_fields.

Did you try the example?  The example should work, even though the files are not there.  (See below.) 
 Incidentally, all of the files you are seeking ("oplsaa.lt", "tip5p.lt", "spce.lt") are located in the
[moltemplate/force_fields/](tree/master/moltemplate/force_fields) subdirectory.

However there is no need to copy these files into your directory.  If moltemplate cannot find the files it needs, it will automatically look in the "moltemplate/force_fields/" directory automatically.  (You can override this and tell moltemplate to look somewhere else by defining a shell environment variable named "MOLTEMPLATE_PATH").    If use this strategy, because it is not possible to copy the "oplsaa.lt" file to every example.  If I did that, the moltemplate download would be quite large.  ("oplsaa.lt" is a big file.)

   Perhaps I should make this more clear in the manual.

> And no the most important tip5p_oplsaa.lt in the folder!(i really need this example. i am doing a simulation about oplsaa with tip4p, but it always "error: out of ranges: Out of range atoms - cannot compute PPPM". 

The "tip5p.lt" file is also located in the "moltemplate/force_fields/" subdirectory.
There is an [example here](tree/master/examples/all_atom/force_field_OPLSAA/waterTIP5P+ethanol)

> after visualize i find dummy atom M suck with other molecule's O together at nve/limit stage at the beginning, when turn to npt error raise. Whether i use fix shake all can not work. Do you have any)
> 
> And: there is no fix fxShake in system.in.settings file. But the only place it appear is in old version spce.lt

The "spce.lt" file is also located in the "moltemplate/force_fields/" subdirectory.  It contains the information needed to run fix_shake, which you showed below:

> ```
> #  ..... some other information......
> #  
> #  write_once("In Settings") {
> #    bond_coeff   @bond:OH     harmonic               1000.0  1.0 
> #    angle_coeff  @angle:HOH   harmonic               1000.0  109.47
> #    pair_coeff   @atom:O @atom:O  lj/cut/coul/long   0.1553  3.166 
> #    pair_coeff   @atom:H @atom:H  lj/cut/coul/long   0.0     2.058
> #    group spce type  @atom:O  @atom:H
> #    fix fShakeSPCE spce shake 0.0001 10 100 b @bond:OH a @angle:HOH
> #    # (Remember to "unfix" fShakeSPCE during minimization.)
> #  }
> #  }
> ```

(This text appears in the "In Settings" section.  That means that the "fix shake" command and the "pair_coeff", "bond_coeff", and "angle_coeff" commands will appear in the "system.in.settings" file after moltemplate.sh is finished.)

> I hope we can get together to make moltemplate great XD.

I am happy and flattered that you have large ambitions for moltemplate and want to improve it.  However I don't think you realize how much work will be required to make moltemplate a useful tool for simulating proteins.  Problem 2 is a big, complicated problem.

The recent OpenFF/SMIRNOFF force field may avoid problem 2 because it does not use atom types (like "@atom:166" or "@atom:c3") to to look up angle, dihedral, and improper interactions.  For this reason, eventually I would like to make OpenFF/SMIRNOFF available in moltemplate.  (I think OpenFF/SMIRNOFF is the future.)  But currently, OpenFF/SMIRNOFF does not work with moltemplate.

Perhaps if we cannot solve problem 2, then at least you can help me do a better job of explaining the current limitations of moltemplate to other LAMMPS users.  I don't want people to spend a lot of time trying to use moltemplate and then realize that it is not useful for their needs.  I prefer to warn them at the beginning:

**Moltemplate does not automatically determine atom type**

Let me know your opinions.  Also let me know if you are confused by anything I said.  It is easy to be confused when discussing these detailed, complicated issues, and I don't mind to explain myself again.

Andrew

**Comment by jewettaij** [comment]
I finally added all of your suggestions to the moltemplate manual.  I hope that helps future users.  Thanks for your suggestions and feedback.  I think I will close this issue now.
